### PR TITLE
Use cluster average skip rate in stake-o-matic

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Always have to adjust a fixed average when skip rate is variable.

#### Summary of Changes

Compare block producers against each other, classify based on a skip percentage below the mean.

Fixes #

### CODE REVIEW

File: stake-o-matic/src/main.rs [hunk 1]
```diff
@@ -562,6 +562,64 @@ where
...
+            quality_block_producers.insert(validator_identity);
+        }
+    }
+
+    info!("quality_block_producers: {}", quality_block_producers.len());
+    trace!("quality_block_producers: {:?}", quality_block_producers);
+    info!("poor_block_producers: {}", poor_block_producers.len());
+    trace!("poor_block_producers: {:?}", poor_block_producers);
+
+    Ok((quality_block_producers, poor_block_producers))
```
Comment (t-nelson) [inline stake-o-matic/src/main.rs:620]: Can we return `cluster_average_rate` as well and notify on it beyond some threshold?  This is for visibility more than anything else. :lobster:

**Review by t-nelson — DISMISSED** [review]
A very welcome change, thanks!  Looks great! Just one nit to try to prevent being caught off guard by a creeping poor cluster average

File: stake-o-matic/src/main.rs (hunk 1 above)
Comment (sakridge) [reply stake-o-matic/src/main.rs:620]: Like just doing a `info!` if it's above 50% or something?

Comment (t-nelson) [reply stake-o-matic/src/main.rs:620]: That might be enough, but I was thinking of returning it back to main and `notifier.send()`'ing it, so we see on Discord

Comment (t-nelson) [reply stake-o-matic/src/main.rs:620]: Around here would be good, https://github.com/solana-labs/solana/blob/24af9961d2cd9acbafca4869535b6c18346d0a0f/stake-o-matic/src/main.rs#L1517-L1523

Comment (sakridge) [reply stake-o-matic/src/main.rs:620]: oh ok, I see.

File: stake-o-matic/src/main.rs [hunk 2]
```diff
@@ -1482,6 +1529,17 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
+    if cluster_average_skip_rate > config.bad_cluster_average_skip_rate {
+        let message = format!(
+            "Cluster average skip rate: {} is above threshold: {}",
+            cluster_average_skip_rate, config.bad_cluster_average_skip_rate
+        );
+        warn!("{}", message);
+        if !config.dry_run {
+            notifier.send(&message);
+        }
+    }
```
Comment (t-nelson) [inline stake-o-matic/src/main.rs:1541]: Perfect!  Thanks!

**Review by t-nelson — APPROVED** [review]

**Comment by felipe-vvoosh** [comment]
Would it be possible for you to explain if there's any logic change in the algorithm to determine if someone is a poor producer or not?

I tried to understand the code but I confess it looks more complex now, just wanted to see in the comment of this PR the changes made...

For example:
If the skip rate of the cluster is higher than the skip rate of the node, this node is a quality producer.
If not, it's a poor producer...

Just wanted to know in human words hehe...

Thanks!

**Comment by meyerbro** [comment]
Would it be possible for you to explain if there's any logic change in the algorithm to determine if someone is a poor producer or not?

I tried to understand the code but I confess it looks more complex now, just wanted to see in the comment of this PR the changes made...

For example:
If the skip rate of the cluster is higher than the skip rate of the node, this node is a quality producer.
If not, it's a poor producer...

Just wanted to know in human words hehe...

Thanks!

**Comment by meyerbro** [comment]
I think the logic before was:
any skiprate above 25% is a poor producer right?

**Comment by t-nelson** [comment]
@meyerbro I think you got an answer for this on Discord, but I'll repeat it here for posterity.

If you consider the skip rate destaking criteria to be

`VALIDATOR_SKIP_RATE - CLUSTER_FLOOR >= THRESHOLD_RATE`

where the original code had a `CLUSTER_FLOOR` of `0`.  This change raises `CLUSTER_FLOOR` to `CLUSTER_AVERAGE_SKIP_RATE`.

Example:
```
VALIDATOR_SKIP_RATE: 20
THRESHOLD_SKIP_RATE: 15
CLUSTER_AVERAGE_SKIP_RATE: 10

// old code
20 - 0 >= 15 // true -> destaked

// new code
20 - 10 >= 15 // false -> stake retained
```

**Pull request merged**